### ISSUE TITLE
Convert native compile/link tasks to be parallel by default

### ISSUE THREAD

**Issue by ghale** [gradle/gradle-native#49]
Use internal APIs to make compile/link tasks safely parallel by default.

Test cases:
- Link executable tasks execute in parallel by default
- Link shared library tasks execute in parallel by default
- Create static library tasks execute in parallel by default

**Comment by ghale** [comment]
PR: https://github.com/gradle/gradle/pull/2628

Note that I'm going to pull making install tasks parallel out into a separate story.  That work is sufficiently different (and this story is big enough as is) that I think it warrants being handled separately.

**Comment by ghale** [comment]
New PR for this: https://github.com/gradle/gradle/pull/2656

Much simpler with @adammurdoch's recommendations.  Note that the problem with #70 needs to be fixed before this should be merged, but the changes for that should not have an effect on these changes at all, so it can be reviewed as is.

**Comment by ghale** [comment]
Also, with the simpler approach, it's actually simpler to just make all compile/link tasks parallel rather than treat them as separate stories.

**Comment by ghale** [comment]
@eljobe This shouldn't be closed - it's still under review.  I'm assuming this was closed due to some confusion around moving issues around and not on purpose, so I'm going to reopen it.

**Comment by ghale** [comment]
This has been merged.

**Comment by adammurdoch** [comment]
Don't forget release notes. This is a really fantastic improvement and our first parallel by default tasks.

**Comment by ThadHouse** [comment]
How do I disable the added parallel support. Upgrading from gradle 4.1 to 4.6 has almost doubled my compile times, and profiling shows most of this is from the parallel compilation. I can't find a way to disable this, and there really needs to be.

**Comment by big-guy** [comment]
@ThadHouse This only restored behavior that was lost in Gradle 4.0-4.1 and refers to parallel compilation/linking _within_ a project.  Native compile/link tasks have always been able to run in parallel from separate projects.  

I suspect you may be actually seeing #502, which is something we've been chipping away at since a performance regression was introduced in 4.3.  You can probably confirm this by building a subset of your build with 4.1 and 4.6 (or 4.7-rc-1) with `--max-workers=1`.  This disables all parallelism essentially. If you still see a large difference, you're probably seeing the same problem.
